File: frontend/src/domain/job/ports/job-repository.ts
/**
 * Job repository port - defines contract for job persistence.
 */

import { JobId } from '../../shared/branded-types';
import { Result } from '../../shared/result';
import { Job } from '../entities/job.entity';
import { JobNotFoundError } from '../errors/job-errors';

export interface JobRepository {
    save(job: Job): Promise<Result<Job, Error>>;

    findById(jobId: JobId): Promise<Result<Job, JobNotFoundError>>;

    findAll(): Promise<Result<ReadonlyArray<Job>, Error>>;

    delete(jobId: JobId): Promise<Result<void, JobNotFoundError>>;
}

File: backend/streamcraft/domain/job/ports/job_repository.py
"""Job repository port - defines contract for job persistence."""

from abc import ABC, abstractmethod
from typing import Sequence

from streamcraft.domain.job.entities.job import Job
from streamcraft.domain.job.errors.job_errors import JobNotFoundError
from streamcraft.domain.shared.branded_types import JobId
from streamcraft.domain.shared.result import Result


class JobRepository(ABC):
    """Port for job persistence operations."""

    @abstractmethod
    def save(self, job: Job) -> Result[Job, Exception]:
        """Save a job."""
        ...

    @abstractmethod
    def find_by_id(self, job_id: JobId) -> Result[Job, JobNotFoundError]:
        """Find a job by ID."""
        ...

    @abstractmethod
    def find_all(self) -> Result[Sequence[Job], Exception]:
        """Find all jobs."""
        ...

    @abstractmethod
    def delete(self, job_id: JobId) -> Result[None, JobNotFoundError]:
        """Delete a job."""
        ...
